include/deal.II/cgal/triangulation.h

```diff
@@ -0,0 +1,259 @@
+    }
+
+    auto has_faces = boost::hana::is_valid(
+      [](auto &&obj) -> decltype(obj.finite_face_handles()) {});
+
+    auto has_cells = boost::hana::is_valid(
+      [](auto &&obj) -> decltype(obj.finite_cell_handles()) {});
+
+    // Different loops for Triangulation_2 and Triangulation_3 types.
+    if constexpr (decltype(has_faces(cgal_triangulation)){})
+      {
+        // This is a non-degenerate Triangulation_2
+        if (cgal_triangulation.dimension() == 2)
+          for (const auto f : cgal_triangulation.finite_face_handles())
+            {
+              CellData<dim> cell(3);
+              for (unsigned int i = 0; i < 3; ++i)
+                cell.vertices[i] = vertex_map[f->vertex(i)];
+              cells.push_back(cell);
+            }
+        else if (cgal_triangulation.dimension() == 1)
+          // This is a degenerate Triangulation_2, made of edges
+          for (const auto e : cgal_triangulation.finite_edges())
+            {
+              // An edge is idenfied by a face and a vertex index in the face
+              const auto &  f = e.first;
+              const auto &  i = e.second;
+              CellData<dim> cell(2);
+              unsigned int  id = 0;
+              for (unsigned int j = 0; j < 3; ++j)
+                if (j != i)
+                  cell.vertices[id++] = vertex_map[f->vertex(j)];
+              cells.push_back(cell);
+            }
+        else
+          {
+            Assert(false, ExcInternalError());
+          }
+      }
+    else if constexpr (decltype(has_cells(cgal_triangulation)){})
+      {
+        // This is a non-degenerate Triangulation_3
+        if (cgal_triangulation.dimension() == 3)
+          for (const auto c : cgal_triangulation.finite_cell_handles())
+            {
+              CellData<dim> cell(4);
+              for (unsigned int i = 0; i < 4; ++i)
+                cell.vertices[i] = vertex_map[c->vertex(i)];
+              cells.push_back(cell);
+            }
+        else if (cgal_triangulation.dimension() == 2)
+          // This is a degenerate Triangulation_3, made of triangles
+          for (const auto facet : cgal_triangulation.finite_facets())
+            {
+              // A facet is idenfied by a cell and the opposite vertex index in
+              // the face
+              const auto &  c = facet.first;
+              const auto &  i = facet.second;
+              CellData<dim> cell(3);
+              unsigned int  id = 0;
+              for (unsigned int j = 0; j < 4; ++j)
+                if (j != i)
+                  cell.vertices[id++] = vertex_map[c->vertex(j)];
+              cells.push_back(cell);
+            }
+        else if (cgal_triangulation.dimension() == 1)
+          // This is a degenerate Triangulation_3, made of edges
+          for (const auto edge : cgal_triangulation.finite_edges())
+            {
+              // An edge is idenfied by a cell and its two vertices
+              const auto &[c, i, j] = edge;
```

Comment:
This is also a C++17 feature, isn't it?

Reply:
Yes. The wrappers are only enabled if c++17 is enabled.